### PR TITLE
RR graph generation from device database

### PR DESCRIPTION
**Pull request by mtdudek**
This PR is WIP of RR graph generation in VPR from FPGA Interchange device database

### CODE REVIEW

File: vpr/src/base/vpr_types.h [hunk 1]
```diff
@@ -1278,6 +1278,9 @@ struct t_router_opts {
     e_rr_node_reorder_algorithm reorder_rr_graph_nodes_algorithm = DONT_REORDER;
     int reorder_rr_graph_nodes_threshold = 0;
     int reorder_rr_graph_nodes_seed = 1;
+
+    // Options related to FPGA INterchange
```
Comment (acomodi) [inline vpr/src/base/vpr_types.h]: Typo
```suggestion
    // Options related to FPGA Interchange
```

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 2]
```diff
@@ -0,0 +1,1259 @@
...
+            } else if (half.getMax().isMax()) {
+                return half.getMax().getMax();
+            } else {
+                VPR_THROW(VPR_ERROR_ARCH, "Invalid speed model %s. No value found!\n", speed_model);
+            }
+        }
+    }
+
+    return 0.;
+}
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]: Not sure where to best put this, but it seems that it is actually repeated from the arch reader. We might have yet another library in `libs/libarchfpga/src/` dedicated to interchange reader helpers.

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 3]
```diff
@@ -0,0 +1,1259 @@
...
+
+typedef std::tuple<int, int> loc_;
+
+/*
+ * Intermediate data type.
+ * It is used to build and explore FPGA Interchange
+ * nodes graph and later convert them to rr_nodes.
+ */
+
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]: remove extra new-line

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 4]
```diff
@@ -0,0 +1,1259 @@
...
+            } else {
+                VPR_THROW(VPR_ERROR_ARCH, "Invalid speed model %s. No value found!\n", speed_model);
+            }
+        }
+    }
+
+    return 0.;
+}
+
+typedef std::tuple<int, int> loc_;
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]: Comment on what this is used for. Also, I think it might be better to have the last `_` only for private variables that are used within structs or classes, to clearly understand that those are internals to the struct.

Using the same methodology for other types/variables can create confusion.

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 5]
```diff
@@ -0,0 +1,1259 @@
...
+
+struct node_ {
+  public:
+    loc_ loc;
+    int segment_id;
+    node_* links[4]; //left, up, right, down
+    node_* visited;
+    bool on_route;
+    bool has_pins;
+    node_() = delete;
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]: Why is this needed?

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 6]
```diff
@@ -0,0 +1,1259 @@
...
+        visited = nullptr;
+        on_route = false;
+        has_pins = false;
+    }
+};
+
+/*
+ * Main class for FPGA Interchange device to rr_graph flow
+ */
+
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]: remove extra new-line

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 7]
```diff
@@ -0,0 +1,1259 @@
...
+  private:
+    Device::Reader& ar_;
+    DeviceContext& device_ctx_;
+    const std::vector<t_segment_inf>& segment_inf_;
+    const DeviceGrid& grid_;
+    const enum e_base_cost_type base_cost_type_;
+
+    vtr::interned_string empty_;
+
+    std::unordered_map<std::string, const t_physical_tile_type*> tile_type_to_pb_type;
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]: add `_` to the name of this variable to identify it as an internal private one.

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 8]
```diff
@@ -0,0 +1,1259 @@
...
+
+    std::unordered_map<std::string, int> wire_name_to_seg_idx_;
+
+    std::set<loc_> empty_tiles;
+
+    std::map<int /*tile_id(name)*/, loc_> tile_to_loc_;
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]: This might be moved to be an `unordered_map`

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 9]
```diff
@@ -0,0 +1,1259 @@
...
+     * Intermediate data storage.
+     * - Shorts represent connections between rr_nodes in a single FPGA Interchange node.
+     * - Redirect is map from node_id at location to location channel and index of track in that channel.
+     * It's useful for ends of the nodes (FPGA Interchange) that don't have representation in CHANS.
+     * - CHAN_loc_map maps from location and CHAN to vector containing tracks description.
+     */
+
+    std::vector<std::tuple<loc_, e_rr_type, int /*idx*/, loc_, e_rr_type, int /*idx*/>> shorts_;
+    std::map<std::tuple<int /*node_id*/, loc_>, std::tuple<loc_, e_rr_type /*CHANX/CHANY*/, int /*idx*/>> redirect_;
+    std::map<std::tuple<loc_, e_rr_type /*CHANX/CHANY*/>, std::vector<std::tuple<int, float, float>> /*idx = ptc,<seg_idx, R, C>*/> CHAN_loc_map_;
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]:
```suggestion
    std::map<std::tuple<loc_, e_rr_type /*CHANX/CHANY*/>, std::vector<std::tuple<int, float, float>> /*idx = ptc,<seg_idx, R, C>*/> chan_loc_map_;
```

I think it is better to leave variable names all in the same style, just to reduce confusion later on where this map is used.

On an additional note, you may consider whether it is worth creating additional structs instead of having tuples, just to improve maintainability. E.g. it is easier to update a struct rather than modify a tuple definition.

Comment (acomodi) [reply vpr/src/route/rr_graph_fpga_interchange.cpp]: In addition, you may also introduce hash functions in a struct and be then able to use `unordered_map` instead of `map` to improve on performance.

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 10]
```diff
@@ -0,0 +1,1259 @@
...
+     */
+
+    std::set<std::tuple<int /*node_id*/, loc_>> used_by_pip_;
+    std::set<std::tuple<int /*node_id*/, loc_>> used_by_pin_;
+
+    /* Sink_source_loc_map is used to create ink/source and ipin/opin rr_nodes,
+     * rr_edges from sink/source to ipin/opin and from ipin/opin to their coresponding segments
+     * uniq_id is used to create edge from segment to ipin/opin, edges from ipin/opin to sink/source are created simply as one to one
+     */
+    std::map<int /*tile_id*/, std::vector<std::tuple<bool, float, int>> /*idx = ptc,<type, R/C, node_id>*/> sink_source_loc_map_;
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]: This could be an `unordered_map`

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 11]
```diff
@@ -0,0 +1,1259 @@
...
+                seen.insert(t1);
+                if (!pip.getDirectional()) {
+                    std::tuple<int, bool> t2{pip.getTiming(), pip.getBuffered20()};
+                    seen.insert(t2);
+                }
+            }
+        }
+        device_ctx_.rr_switch_inf.reserve(seen.size() + 1);
+        int id = 2;
+        make_room_in_vector(&device_ctx_.rr_switch_inf, 0);
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp:331]: Out of curiosity, where is this `make_room_in_vector` defined?

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 12]
```diff
@@ -0,0 +1,1259 @@
...
+        device_ctx_.rr_switch_inf[0].R = 0;
+        device_ctx_.rr_switch_inf[0].Cin = 0;
+        device_ctx_.rr_switch_inf[0].Cout = 0;
+        device_ctx_.rr_switch_inf[0].Cinternal = 0;
+        device_ctx_.rr_switch_inf[0].Tdel = 0;
+        device_ctx_.rr_switch_inf[0].buf_size = 0;
+        device_ctx_.rr_switch_inf[0].mux_trans_size = 0;
+        device_ctx_.rr_switch_inf[0].name = "short";
+        device_ctx_.rr_switch_inf[0].set_type(SwitchType::SHORT);
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp:340]: We may consider having a utility function to generate an empty switch to which we just pass the name and type, and optionally the rest of the values.

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 13]
```diff
@@ -0,0 +1,1259 @@
...
+                    if (!pip.getBuffered20()) {
+                        source_sink = std::tuple<int, int>(node0, node1);
+                        pips_[source_sink] = std::tuple<int, std::tuple<int, int, int, bool>>{switch_id, {name, wire0_name, wire1_name, true}};
+                    }
+                }
+            }
+        }
+    }
+
+    loc_ offsets[4] = {loc_(-1, 0), loc_(0, 1), loc_(1, 0), loc_(0, -1)};
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]: This seems to be out of place, possibly it should be defined before the struct's functions

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 14]
```diff
@@ -0,0 +1,1259 @@
...
+    /*
+     * Build graph of FPGA Interchange node for further computations
+     */
+
+    node_* build_node_graph(int node_id,
+                            std::set<loc_> nodes,
+                            std::set<loc_> empties,
+                            int root_x,
+                            int root_y) {
+        int cnt = 0;
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]: I am not sure how this `cnt` represents actually, some description around it may be useful, and possibly also rename it to something more descriptive.

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 15]
```diff
@@ -0,0 +1,1259 @@
...
+            } else if (vertex->on_route) {
+                vertex->visited->on_route = true;
+                if (vertex->visited->segment_id == -1)
+                    vertex->visited->segment_id = vertex->segment_id;
+            }
+        }
+        return cnt;
+    }
+
+    void populate_CHAN_loc_map(int node_id, node_* root, float R, float C) {
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]:
```suggestion
    void populate_chan_loc_map(int node_id, node_* root, float R, float C) {
```

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 16]
```diff
@@ -0,0 +1,1259 @@
...
+        /*
+         * Create future rr_nodes that corespond to given FPGA Interchange node
+         */
+
+        std::map<std::tuple<loc_, e_rr_type /*CHANX/CHANY*/>, int /*index*/> loc_chan_idx;
+
+        std::queue<node_*> walker;
+        std::stack<node_*> back_walker;
+        walker.push(root);
+        bool all_nulls, CHANX, CHANY;
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]:
```suggestion
        bool all_nulls, is_chanx, is_chany;
```

To better understand that this is a boolean varialbe it is good to add some more descriptive name

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 17]
```diff
@@ -0,0 +1,1259 @@
...
+                                                                                         std::get<1>(key1),
+                                                                                         loc_chan_idx[key1],
+                                                                                         std::get<0>(key2),
+                                                                                         std::get<1>(key2),
+                                                                                         loc_chan_idx[key2]));
+            }
+        }
+
+        /*
+         * Create edges between future rr_nodes that corespond to given FPGA Interchange node
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]:
```suggestion
         * Create edges between future rr_nodes that correspond to the given FPGA Interchange node
```

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 18]
```diff
@@ -0,0 +1,1259 @@
...
+                            loc, e_rr_type::CHANY, loc_chan_idx[std::tuple<loc_, e_rr_type>(loc, e_rr_type::CHANY)]});
+                }
+            }
+        }
+    }
+
+    /*
+     * Remove a graph represented by a root node.
+     */
+
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]: This comment might be enhanced with a description on why we should remove the graph

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 19]
```diff
@@ -0,0 +1,1259 @@
...
+            float capacitance, resistance;
+            auto model = ar_.getNodeTimings()[node.getNodeTiming()];
+            capacitance = get_corner_value(model.getCapacitance(), "slow", "typ") / div;
+            resistance = get_corner_value(model.getResistance(), "slow", "typ") / div;
+            populate_CHAN_loc_map(node_id, root, resistance, capacitance);
+            delete_nodes(root);
+        }
+    }
+
+    void pack_chan_width() {
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp:918]: This needs a comment on what id does and why

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 20]
```diff
@@ -0,0 +1,1259 @@
...
+        pack_chan_width();
+        // switches are already packed
+        // segments are already packed before rr_generation
+        // physical_tile_types are already packed before rr_generation
+        // grid is already packed before rr_generation
+        pack_rr_nodes();
+        pack_rr_edges();
+    }
+
+    void finish() {
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]: Might use a more descriptive name for this function, e.g. what is it finishing?

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 21]
```diff
@@ -0,0 +1,1259 @@
...
+        for (size_t i = 0; i < seg_index_.size(); ++i) {
+            device_ctx_.rr_indexed_data[RRIndexedDataId(i)].seg_index = seg_index_[RRIndexedDataId(i)];
+        }
+    }
+
+    /*
+     * Create site sink/source ipin/opin mappins.
+     */
+
+    void create_sink_source_nodes() {
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]: I believe that this function might be moved on top with the other functions that prepare the data for the rr_graph_builder.

**Review by acomodi — CHANGES_REQUESTED** [review]
It looks pretty good. I have left some comments for a first review pass.

I think that it's worth having a general top-level comment on how the RR graph builder works.

File: vpr/src/route/rr_graph_fpga_interchange.cpp (hunk 2 above)
Comment (mtdudek) [reply vpr/src/route/rr_graph_fpga_interchange.cpp]: Sounds good, as it will help us remove repeating code and potential bug repetitions

File: libs/libarchfpga/src/read_fpga_interchange_arch.cpp [hunk 22]
```diff
@@ -1533,7 +1533,7 @@ struct ArchReader {
         //        the RR graph generation is correct.
         //        This can be removed once the RR graph reader from the interchange
         //        device is ready and functional.
-        int num_seg = 1; //wire_names.size();
+        int num_seg = wire_names.size();
```
Comment (acomodi) [inline libs/libarchfpga/src/read_fpga_interchange_arch.cpp:1457]: I believe that also the `if (index >= num_seg) break;` line below should be removed, right?

File: vpr/src/route/rr_graph_fpga_interchange.cpp (hunk 11 above)
Comment (mtdudek) [reply vpr/src/route/rr_graph_fpga_interchange.cpp:331]: It's defined in vpr/src/util/vpr_utils.h

File: vpr/src/route/rr_graph_fpga_interchange.cpp (hunk 21 above)
Comment (mtdudek) [reply vpr/src/route/rr_graph_fpga_interchange.cpp]: fixed in new commit

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 23]
```diff
@@ -0,0 +1,1179 @@
...
+/*
+ * Intermediate data type.
+ * It is used to build and explore FPGA Interchange
+ * nodes graph and later convert them to rr_nodes.
+ */
+struct node_ {
+  public:
+    location loc;
+    int segment_id;
+    node_* links[4]; //left, up, right, down
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]: I think it might be worth having an `enum` class with `LEFT, TOP, RIGHT, BOTTOM` and instead of a C-style array, having a map or vector that is indexed with the enum values, so we can access this by doing:
```
links[UP] = ...
```

I think it would be clearer this way

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 24]
```diff
@@ -0,0 +1,1179 @@
...
+#include "check_rr_graph.h"
+#include "rr_metadata.h"
+#include "rr_graph_indexed_data.h"
+#include "fpga_interchange_arch_utils.h"
+
+using namespace DeviceResources;
+using namespace LogicalNetlist;
+using namespace capnp;
+
+typedef std::tuple<int, int> location;
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]: Having a `std::pair` for types that only contain two simple types is better as it simplify the access and, instead of

```
std::get<0>(loc)
```

you can simply have:

```
loc.first
```

This would improve readability throughout the code.

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 25]
```diff
@@ -0,0 +1,1179 @@
...
+    std::map<std::tuple<int /*node_id*/, int /*node_id*/>, std::tuple<int /*switch_id*/, std::tuple<int /*tile_id*/, int /*wire0*/, int /*wire1*/, bool /*forward*/>>> pips_;
+
+    std::map<std::tuple<int, int>, int> wire_to_node_;
+    std::unordered_map<int, std::set<location>> node_to_locs_;
+    std::map<std::tuple<int, int> /*<node, tile_id>*/, int /*segment type*/> node_tile_to_segment_;
+
+    /*
+     * Offsets for FPGA Interchange node processing
+     */
+    location offsets[4] = {location(-1, 0), location(0, 1), location(1, 0), location(0, -1)};
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp:178]: For clarity it might be good to explicitly have a map or vector (with indices using the above mentioned enum)

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 26]
```diff
@@ -0,0 +1,1179 @@
...
+    int rr_idx = 0; // Do not decrement!
+
+    std::string str(int idx) {
+        return std::string(ar_.getStrList()[idx].cStr());
+    }
+
+    /*
+     * Debug print function
+     */
+    void print_() {
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]: remove final `_`

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 27]
```diff
@@ -0,0 +1,1179 @@
...
+        }
+        return root_node;
+    }
+
+    /*
+     * Removes dangling nodes from a graph represented by the root node.
+     * Dangling nodes are nodes that do not connect to a pin, a pip or other non-dangling node.
+     */
+    int reduce_graph_and_count_nodes(node_* root) {
+        int cnt = 0;
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp:554]: Rename this with a more descriptive name

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 28]
```diff
@@ -0,0 +1,1179 @@
...
+            if (is_chanx && is_chany) {
+                shorts_.emplace_back(std::get<0>(key1), std::get<1>(key1), loc_chan_idx[key1],
+                                     std::get<0>(key2), std::get<1>(key2), loc_chan_idx[key2]);
+            }
+        }
+
+        /*
+         * Create edges between future rr_nodes that correspond to the given FPGA Interchange node
+         */
+        while (!back_walker.empty()) {
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp:664]: In the body of this while, replace all `links[<interger index>]` with `links[<enum value>]`

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 29]
```diff
@@ -0,0 +1,1179 @@
...
+            }
+        }
+        while (!back_walker.empty()) {
+            node_* vertex = back_walker.top();
+            back_walker.pop();
+            for (int i = 0; i < 4; ++i) {
+                node_* temp = nullptr;
+                if (vertex->links[i] != nullptr) {
+                    temp = vertex->links[i];
+                    temp->links[(i + 2) % 4] = nullptr;
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp:804]: Having this equation here might be confusing, probably worth making this explicit to better understand how this is calculated and what does it represent.

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 30]
```diff
@@ -0,0 +1,1179 @@
...
+     */
+    void process_nodes() {
+        auto wires = ar_.getWires();
+        for (auto const& node : ar_.getNodes()) {
+            std::tuple<int, int> base_wire_{wires[node.getWires()[0]].getTile(), wires[node.getWires()[0]].getWire()};
+            int node_id = wire_to_node_[base_wire_];
+            std::set<location> all_possible_tiles = node_to_locs_[node_id];
+            int root_x, root_y;
+            std::tie(root_x, root_y) = *all_possible_tiles.begin();
+            std::set<location> empties;
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]: Worth adding a comment explaining what empties is and why they are needed, just for future reference.

**Review by acomodi — CHANGES_REQUESTED** [review]
Added comments from a second review pass

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 31]
```diff
@@ -0,0 +1,1179 @@
...
+using namespace capnp;
+
+typedef std::tuple<int, int> location;
+
+/*
+ * Intermediate data type.
+ * It is used to build and explore FPGA Interchange
+ * nodes graph and later convert them to rr_nodes.
+ */
+struct node_ {
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]: This may need to be renamed to something else, just to avoid confusion on which kind of node it represents. Maybe something like `node_data` or `intermediate_node`

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 32]
```diff
@@ -0,0 +1,1179 @@
...
+
+            } else if (loc_chan_idx.find(std::make_tuple(loc, e_rr_type::CHANY)) != loc_chan_idx.end()) {
+                redirect_.emplace(std::make_tuple(node_id, loc), std::make_tuple(
+                                                                     loc, e_rr_type::CHANY, loc_chan_idx[std::make_tuple(loc, e_rr_type::CHANY)]));
+
+            } else if (vertex->links[2] != nullptr && (vertex->links[2]->visited == vertex || vertex->links[2] == vertex->visited) && loc_chan_idx.find(std::make_tuple(right, e_rr_type::CHANX)) != loc_chan_idx.end()) {
+                redirect_.emplace(std::make_tuple(node_id, loc), std::make_tuple(
+                                                                     right, e_rr_type::CHANX, loc_chan_idx[std::make_tuple(right, e_rr_type::CHANX)]));
+
+            } else if (vertex->links[3] != nullptr && (vertex->links[3]->visited == vertex || vertex->links[3] == vertex->visited) && loc_chan_idx.find(std::make_tuple(down, e_rr_type::CHANY)) != loc_chan_idx.end()) {
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]: In general, it is better to split and pre-compute a very long `if` condition to make it clearer.

File: vpr/src/route/rr_graph_fpga_interchange.cpp [hunk 33]
```diff
@@ -19,25 +19,35 @@ using namespace DeviceResources;
 using namespace LogicalNetlist;
 using namespace capnp;
 
-typedef std::tuple<int, int> location;
+typedef std::pair<int, int> location;
+
+enum g_sides {
```
Comment (acomodi) [inline vpr/src/route/rr_graph_fpga_interchange.cpp]: what does `g` stand for?

Comment (mtdudek) [reply vpr/src/route/rr_graph_fpga_interchange.cpp]: g -> graph
yeah I should came up with better name

File: vpr/src/route/rr_graph_fpga_interchange.cpp (hunk 27 above)
Comment (mtdudek) [reply vpr/src/route/rr_graph_fpga_interchange.cpp:554]: I hope this is good enough

**Comment by mtdudek** [comment]
Superseded by  Mdudek/fpga interchange rr graph gen squashed #10
